ffcx/codegeneration/integrals.py

```diff
@@ -586,92 +588,122 @@ def generate_block_parts(self, quadrature_rule, blockmap, blockdata):
-                continue
-            break
-
-        if expand_loop:
-            # If DOFs in dofrange are not equally spaced, then expand
-            # out the for loop
-            for A_indices, B_indices in zip(itertools.product(*blockmap),
-                                            itertools.product(*[range(len(b)) for b in blockmap])):
-                quadparts += [
-                    L.AssignAdd(
-                        A[A_indices],
-                        L.float_product([fw] + self.get_arg_factors(
-                            blockdata, block_rank,
-                            quadrature_rule, iq, B_indices)
-                        )
-                    )
-                ]
-        else:
+                # Define and cache scalar temp variable
+                key = (quadrature_rule, factor_index, blockdata.all_factors_piecewise)
+                fw, defined = self.get_temp_symbol("fw", key)
+                if not defined:
+                    quadparts.append(L.VariableDecl("const ufc_scalar_t", fw, fw_rhs))
+
+            assert not blockdata.transposed
+            A_shape = self.ir.tensor_shape
+
+            Asym = self.backend.symbols.element_tensor()
+            A = L.FlattenedArray(Asym, dims=A_shape)
+
             # Fetch code to access modified arguments
             arg_factors = self.get_arg_factors(blockdata, block_rank, quadrature_rule, iq, B_indices)
 
             B_rhs = L.float_product([fw] + arg_factors)
-            A_indices = []
 
-            for bm, index in zip(blockmap, arg_indices):
-                # TODO: switch order here? (optionally)
-                offset = bm[0]
-                if len(bm) == 1:
-                    A_indices.append(index + offset)
-                else:
-                    block_size = bm[1] - bm[0]
-                    A_indices.append(block_size * index + offset)
+            # Accumulate RHS on temporary scalar
+            rhs_list.append(B_rhs)
 
-            body = L.AssignAdd(A[A_indices], B_rhs)
-
-            for i in reversed(range(block_rank)):
-                body = L.ForRange(B_indices[i], 0, blockdims[i], body=body)
-            quadparts += [body]
+        A_indices = []
+        for bm, index in zip(blockmap, arg_indices):
+            # TODO: switch order here? (optionally)
+            offset = bm[0]
+            if len(bm) == 1:
+                A_indices.append(index + offset)
+            else:
+                block_size = bm[1] - bm[0]
+                A_indices.append(block_size * index + offset)
+
+        hoist_rhs = collections.defaultdict(list)
+        # List of statements to keep in the inner loop
+        keep = []
+        # List of temporary array declarations
+        pre_loop = []
+        # List of loop invariant expressions to hoist
+        hoist = []
+
+        # Hoist loop invariant code and group array access (each table should only be read one
+        # time in the inner loop).
+        if block_rank == 2:
```

Comment:
Are these optimisations not applicable to vector assembly, `block_rank == 1`?

Reply:
Not in the current form.
In vector assembly, we have in general `fwX * TableY[i];` in the inner loop; so this optimization would have no effect.

But we are planning to conduct a more extensive analysis of the performance of the vector assembly.

Reply:
I had blocks loop fusion in mind, but I see now this is applied also for vector assembly.